### DISCUSSION TITLE
Special Tokens

### DISCUSSION THREAD

**Discussion by argosopentech** [argosopentech/argos-translate#65] [Ideas]
## Define
```
<define> procrastinate
To defer or delay action
```

## Translating to many languages with one model
```
<translate-fr> I walked to town to get bread.
Je suis allé en ville chercher du pain.

<translate-ga> I walked to town to get bread.
Shiúil mé go baile a fháil arán.
```

## Overview
Special tokens are tokens that are tokenized as a single token by the tokenizer and add functionality to models. For the \<define\> token this lets us train a model on dictionary data to better understand a language. For multi-translate it allows one (possibly larger) model to translate to multiple languages.

## Current state
Define tokens are implemented and Wiktionary data is currently being trained using them. The current models aren't able to define words but hopefully the extra data is improving translation quality. Multi-translation isn't currently being done but I think fine tuning a single pretrained language model to translate many languages is going to be a better architecture in the future.

## References
- [Stanford Seminar: Google's Multilingual Neural Machine Translation System](https://youtu.be/nR74lBO5M3s?t=1682)

**Comment by PJ-Finlay** [comment]
I have a model [doing sentence boundary detection](https://github.com/argosopentech/onmt-models/tree/0e1e317a752792b1ad9a73d986d30dbd9c054617/sbd) using the seq2seq model and special tokens that seems to be working reasonably well. I'm going to try combining it with a translation model to see if it still works.
